src/lib/adapters/WebDav.js

```diff
@@ -0,0 +1,654 @@
+		{
+			console.log ("obtainLock: 005 " + rStatus);
+			let fullURL = this.getBookmarkLockURL ();
+			console.log (fullURL);
+            this.uploadFile (fullURL, 'text/html', '<html><body>I am a lock file</body></html>');
+		}
+		else {
+			console.log ("obtainLock: 006 " + rStatus);
+        	throw new Error('Network Error: Unable to determine status of lock file ' + this.server.bookmark_file + '.lock');
+		}
+
+        return 1;
+    }
+
+    async freeLock () {
+        console.log ("freeLock: 001");
+
+        let fullUrl = this.server.bookmark_file;
+        fullUrl = this.server.url + fullUrl + ".lock";
+
+        console.log ("fullURL :" + fullUrl + ":");
+
+        let rStatus = 500;
+        let response;
+
+        try {
+            response = await fetch (fullUrl, {
+                    method: 'DELETE',
+                    headers: {
+                        'Authorization': 'Basic ' + btoa(this.server.username + ':' + this.server.password)
+                    },
+                });
+
+            rStatus = response.status;
+
+            console.log ("response");
+            console.log (response);
+        } catch (e) {
+            console.log ("Error Caught");
+            console.log (e);
+        }
+    }
+
+    htmlEncode ( content ) {
+        return document.createElement('a').appendChild(
+            document.createTextNode (content)).parentNode.innerHTML;
+    }
+
+    htmlDecode ( content ) {
+        let a = document.createElement ('a');
+        a.innerHTML = html;
+        return a.textContent;
+    }
+
+	outputFolderXBEL (myStructure, indent) {
+		let output = "";
+
+		myStructure.bookmarks.forEach ((bm) => {
+			output += indent + '<bookmark href=';
+            output += '"' + this.htmlEncode (bm.url) + '"';
+            output += ' id="' + bm.id + `">
+`;	
+			output += indent + "    <title>" + this.htmlEncode (bm.title) + `</title>
+`;
+			output += indent + `</bookmark>
+`;
+		});
+
+		let keys = Object.keys (myStructure.folders);
+		let values = keys.map ((v) => { return myStructure.folders[v]; });
+
+		values.forEach ((folder) => {
+			output += indent + `<folder>
+`;
+			output += indent + '    <title>' + this.htmlEncode (folder.title) + `</title>
+`;
+
+			output += this.outputFolderXBEL (folder, indent + '    ');
+
+			output += indent + `</folder>
+`;
+		});
+
+		return output;
+	}
+
+    createXBEL (myStructure) {
+        
+        let output = `<?xml version="1.0" encoding="ISO-8859-1"?>
+<!DOCTYPE xbel PUBLIC "+//IDN python.org//DTD XML Bookmark Exchange Language 1.0//EN//XML" "http://www.python.org/topics/xml/dtds/xbel-1.0.dtd">
+<xbel version="1.0">
+`;
+
+        output += '<!--- highestID :' + this.highestID + `: for Floccus bookmark sync browser extension -->
+`;
+
+		output += this.outputFolderXBEL (myStructure, '');
+
+		output += `
+</xbel>`;
+
+		return output;
+    }
+
+    /* private routine */
+    _addBookmark (myStructure, bm)
+    {
+        let myArray = bm.path.split ("/");
```

Comment:
This won't work for all folder names, sadly. If they contain slashes, those are escaped with a backslash. Maybe we can find a better solution than to pass the path as a string. ([this](https://github.com/marcelklehr/floccus/blob/master/src/lib/Tree.js#L198) is how I parse those paths, theh)